xkcd/pkg/types/types.go

```diff
@@ -0,0 +1,164 @@
+package types
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+	"strings"
+)
+
+type OndiskSerialization int
+
+const (
+	Protobuf OndiskSerialization = 1 + iota
+	JSON
+)
+
+type OperationType int
+
+const (
+	Update OperationType = 1 + iota
+	Search
+	List
+)
+
+func (s OndiskSerialization) String() string {
+	switch s {
+	case Protobuf:
+		return "protobuf"
+	case JSON:
+		return "json"
+	default:
+		return "unknown"
+	}
+}
+
+func (s *OndiskSerialization) Set(in string) error {
+	switch strings.ToLower(in) {
+	case "protobuf":
+		*s = Protobuf
+	case "json":
+		*s = JSON
+	default:
+		return fmt.Errorf("unrecognized serialization type `%s`", in)
+	}
+	return nil
+}
+
+func (s OperationType) String() string {
+	switch s {
+	case Update:
+		return "update"
+	case Search:
+		return "search"
+	case List:
+		return "list"
+	default:
+		return "unknown"
+	}
+}
+
+func (s *OperationType) Set(in string) error {
+	switch strings.ToLower(in) {
+	case "update":
+		*s = Update
+	case "search":
+		*s = Search
+	case "list":
+		*s = List
+	default:
+		return fmt.Errorf("unrecognized operation `%s`", in)
+	}
+	return nil
+}
+
+// IndexFile bundles together all the parameters describing an index.
+type IndexFile struct {
+	Type     OndiskSerialization
+	Location string
+}
+
+func (r IndexFile) String() string {
+	return fmt.Sprintf("type: %s, location: %s", r.Type, r.Location)
+}
+
+type Range struct {
+	Min, Max int
+}
+
+func (r Range) String() string {
+	return fmt.Sprintf("min: %d, max: %d", r.Min, r.Max)
+}
+
+type Story struct {
+	Alt        string
+	Day        int `json:",string"`
+	Img        string
+	Link       string
+	Month      int `json:",string"`
+	News       string
+	Num        int
+	SafeTitle  string `json:"safe_title"`
+	Title      string
+	Transcript string
+	Year       int `json:",string"`
+}
+
+func (s Story) String() string {
+	res := ""
+
+	res += fmt.Sprintf("\talt: %s\n", s.Alt)
+	res += fmt.Sprintf("\tday: %d\n", s.Day)
+	res += fmt.Sprintf("\timg: %s\n", s.Img)
+	res += fmt.Sprintf("\tlink: %s\n", s.Link)
+	res += fmt.Sprintf("\tmonth: %d\n", s.Month)
+	res += fmt.Sprintf("\tnews: %s\n", s.News)
+	res += fmt.Sprintf("\tnum: %d\n", s.Num)
+	res += fmt.Sprintf("\tsafe_title: %s\n", s.SafeTitle)
+	res += fmt.Sprintf("\ttitle: %s\n", s.Title)
+	res += fmt.Sprintf("\ttranscript: %s\n", s.Transcript)
+	res += fmt.Sprintf("\tyear: %d\n", s.Year)
+
+	return res
+}
+
+type AllStories map[int]*Story
+
+func (a AllStories) String() string {
+	res := ""
+
+	for i := range a {
+		res += fmt.Sprintf("Story %d:\n%s", i, *a[i])
+	}
+
+	return res
+}
+
+func (a AllStories) MarshalJSON() ([]byte, error) {
+	res := make(map[string]Story)
+
+	for k, v := range a {
+		res[strconv.Itoa(k)] = *v
```

Comment:
Couple of things here:
* ACK when it comes to Marshaler - did not know about that.
* my intention is to keep MarshalJson and UnmarshalJson symmetric. Json marshaller taking care of references in this case is a bit of Magic. If you have a person that does not know that, he/she starts wondering why we do not dereference here. I believe that Python's "Explicit is better than implicit." describes the issue perfectly :)
* the *main* reason why I introduced this function (apart from seeing how they work :D) is that AllStories uses `int` as map keys, while Json by default uses strings. These two functions convert between both formats.